torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
+                           dst_param_groups: List[Dict[Any, Any]]) -> None:
         r"""
-        Returns the local_state_dict for a given rank.
+        Syncs the attributes from the source parameter groups to the
+        destination parameter groups.
+
+        Example attributes include learning rate or scheduler attributes. The
+        two parameter groups should have the same length (i.e. same number of
+        parameter groups).
 
         Arguments:
-            rank (int): rank to get ``local_state_dict`` for
-            state_dict (dict): global ``state_dict``
+            src_param_groups (list[dict]): parameter groups giving the
+                attribute settings to copy.
+            dst_param_groups (list[dict]): parameter groups giving the
+                attribute settings to set.
         """
-        param_groups = state_dict["param_groups"][state_dict["partition"][rank][0] : state_dict["partition"][rank][1]]
-        return {"state": state_dict["state"][rank], "param_groups": param_groups}
+        assert len(src_param_groups) == len(dst_param_groups)
+        for src_param_group, dst_param_group in zip(src_param_groups, dst_param_groups):
+            # Sync all attributes except the parameters
+            for attr in filter(lambda x: x != "params", src_param_group.keys()):
+                dst_param_group[attr] = src_param_group[attr]
 
-    @staticmethod
-    def _sync_param_groups(source: List[Dict[Any, Any]], destination: List[Dict[Any, Any]]) -> None:
-        r"""Sync learning rate and other optimizer attributes (needed to support schedulers)."""
+    def _build_param_buckets(self) -> None:
+        r"""
+        Builds parameter buckets so that for each device that stores this
+        rank's parameters, there is a bucket (represented as a tensor)
+        containing all of the parameters on that device that are assigned to a
+        given rank.
 
-        for source_group, destination_group in zip(source, destination):
-            # Sync everything but the parameters
-            for k in filter(lambda x: x != "params", source_group.keys()):
-                destination_group[k] = source_group[k]
+        This function is called in the constructor and any time parameter
+        trainability is changed if ``parameters_as_bucket_view`` is enabled.
 
-    def _setup_flat_buffers(self) -> None:
-        r"""
-        Make all params which are on the same device and tied to the same rank
-        views of a single buffer. This is used at construction time, and
-        anytime parameter trainability is changed (frozen or unfrozen) and
-        ``_update_trainable`` is called.
+        NOTE: The current implementation assumes that each rank stores all of
+        its parameters (i.e. ``self._all_params``) on a single device. This
+        means that there should be exactly ``world_size``-many buckets.
+
+        NOTE: The current implementation assumes that all of the parameters in
```

Comment:
could be worth it adding an assert with a nice error message ? not super likely to be triggered, but in case that happens it would help debugging a fair bit

Reply:
The assumptions in the `NOTE: ...`s should be verified at the very beginning of the constructor (line 171). I was intending the notes for documentation purposes (for a user reading the source or for a developer planning to modify the implementation). 

Do you think we should do a second pass of verifications each time this function `_build_param_buckets()` is called?

Reply:
sorry, a bit late to react, I meant checking for this assumption explicitly (at least the first time the buckets are built ?) and nicely error out if any issue, but maybe that it was already there indeed. Non blocking, just a brain dump